File: ui/src/core/model/resources.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import { toU32 } from '@/lib/number';

export class ResourcesImpl implements Resources {
  public readonly food: number;
  public readonly iron: number;
  public readonly stone: number;
  public readonly wood: number;

  private constructor(resources: Resources) {
    this.food = toU32(resources.food);
    this.iron = toU32(resources.iron);
    this.stone = toU32(resources.stone);
    this.wood = toU32(resources.wood);
  }

  public add(value: number | PartialNullish<Resources>) {
    if (typeof value === 'number') {
      return ResourcesImpl.create({
        food: this.food + value,
        iron: this.iron + value,
        stone: this.stone + value,
        wood: this.wood + value,
      });
    }

    return ResourcesImpl.create({
      food: this.food + (value.food ?? 0),
      iron: this.iron + (value.iron ?? 0),
      stone: this.stone + (value.stone ?? 0),
      wood: this.wood + (value.wood ?? 0),
    });
  }

  public sub(value: number | PartialNullish<Resources>) {
    if (typeof value === 'number') {
      return ResourcesImpl.create({
        food: this.food - value,
        iron: this.iron - value,
        stone: this.stone - value,
        wood: this.wood - value,
      });
    }

    return ResourcesImpl.create({
      food: this.food - (value.food ?? 0),
      iron: this.iron - (value.iron ?? 0),
      stone: this.stone - (value.stone ?? 0),
      wood: this.wood - (value.wood ?? 0),
    });
  }

  public mul(value: number | PartialNullish<Resources>) {
    if (typeof value === 'number') {
      return ResourcesImpl.create({
        food: this.food * value,
        iron: this.iron * value,
        stone: this.stone * value,
        wood: this.wood * value,
      });
    }

    return ResourcesImpl.create({
      food: this.food * (value.food ?? 0),
      iron: this.iron * (value.iron ?? 0),
      stone: this.stone * (value.stone ?? 0),
      wood: this.wood * (value.wood ?? 0),
    });
  }

  public div(value: number | PartialNullish<Resources>) {
    if (typeof value === 'number') {
      return ResourcesImpl.create({
        food: this.food / value,
        iron: this.iron / value,
        stone: this.stone / value,
        wood: this.wood / value,
      });
    }

    return ResourcesImpl.create({
      food: this.food / (value.food ?? 1),
      iron: this.iron / (value.iron ?? 1),
      stone: this.stone / (value.stone ?? 1),
      wood: this.wood / (value.wood ?? 1),
    });
  }

  public has(resources: PartialNullish<Resources>) {
    return (
      this.food >= (resources.food ?? 0) &&
      this.iron >= (resources.iron ?? 0) &&
      this.stone >= (resources.stone ?? 0) &&
      this.wood >= (resources.wood ?? 0)
    );
  }

  public static create(resources?: PartialNullish<Resources>) {
    if (resources instanceof ResourcesImpl) {
      return resources;
    }

    return new ResourcesImpl({
      food: resources?.food ?? 0,
      iron: resources?.iron ?? 0,
      stone: resources?.stone ?? 0,
      wood: resources?.wood ?? 0,
    });
  }

  public static zero() {
    return ResourcesImpl.splat(0);
  }

  public static splat(value: number) {
    return ResourcesImpl.create({
      food: value,
      iron: value,
      stone: value,
      wood: value,
    });
  }
}

File: crates/nil-client/src/client/cheat/resources.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use crate::client::Client;
use crate::error::Result;
use crate::http;
use nil_core::resources::Resources;
use nil_payload::cheat::resources::*;

impl Client {
  pub async fn cheat_get_resources(&self, req: CheatGetResourcesRequest) -> Result<Resources> {
    http::json_post("cheat-get-resources")
      .body(req)
      .server(self.server)
      .maybe_authorization(self.authorization.as_ref())
      .send()
      .await
  }

  pub async fn cheat_set_food(&self, req: CheatSetFoodRequest) -> Result<()> {
    http::post("cheat-set-food")
      .body(req)
      .server(self.server)
      .maybe_authorization(self.authorization.as_ref())
      .send()
      .await
  }

  pub async fn cheat_set_iron(&self, req: CheatSetIronRequest) -> Result<()> {
    http::post("cheat-set-iron")
      .body(req)
      .server(self.server)
      .maybe_authorization(self.authorization.as_ref())
      .send()
      .await
  }

  pub async fn cheat_set_max_food(&self, req: CheatSetMaxFoodRequest) -> Result<()> {
    http::post("cheat-set-max-food")
      .body(req)
      .server(self.server)
      .maybe_authorization(self.authorization.as_ref())
      .send()
      .await
  }

  pub async fn cheat_set_max_iron(&self, req: CheatSetMaxIronRequest) -> Result<()> {
    http::post("cheat-set-max-iron")
      .body(req)
      .server(self.server)
      .maybe_authorization(self.authorization.as_ref())
      .send()
      .await
  }

  pub async fn cheat_set_max_resources(&self, req: CheatSetMaxResourcesRequest) -> Result<()> {
    http::post("cheat-set-max-resources")
      .body(req)
      .server(self.server)
      .maybe_authorization(self.authorization.as_ref())
      .send()
      .await
  }

  pub async fn cheat_set_max_silo_resources(
    &self,
    req: CheatSetMaxSiloResourcesRequest,
  ) -> Result<()> {
    http::post("cheat-set-max-silo-resources")
      .body(req)
      .server(self.server)
      .maybe_authorization(self.authorization.as_ref())
      .send()
      .await
  }

  pub async fn cheat_set_max_stone(&self, req: CheatSetMaxStoneRequest) -> Result<()> {
    http::post("cheat-set-max-stone")
      .body(req)
      .server(self.server)
      .maybe_authorization(self.authorization.as_ref())
      .send()
      .await
  }

  pub async fn cheat_set_max_warehouse_resources(
    &self,
    req: CheatSetMaxWarehouseResourcesRequest,
  ) -> Result<()> {
    http::post("cheat-set-max-warehouse-resources")
      .body(req)
      .server(self.server)
      .maybe_authorization(self.authorization.as_ref())
      .send()
      .await
  }

  pub async fn cheat_set_max_wood(&self, req: CheatSetMaxWoodRequest) -> Result<()> {
    http::post("cheat-set-max-wood")
      .body(req)
      .server(self.server)
      .maybe_authorization(self.authorization.as_ref())
      .send()
      .await
  }

  pub async fn cheat_set_resources(&self, req: CheatSetResourcesRequest) -> Result<()> {
    http::post("cheat-set-resources")
      .body(req)
      .server(self.server)
      .maybe_authorization(self.authorization.as_ref())
      .send()
      .await
  }

  pub async fn cheat_set_stone(&self, req: CheatSetStoneRequest) -> Result<()> {
    http::post("cheat-set-stone")
      .body(req)
      .server(self.server)
      .maybe_authorization(self.authorization.as_ref())
      .send()
      .await
  }

  pub async fn cheat_set_wood(&self, req: CheatSetWoodRequest) -> Result<()> {
    http::post("cheat-set-wood")
      .body(req)
      .server(self.server)
      .maybe_authorization(self.authorization.as_ref())
      .send()
      .await
  }
}
